### PR TITLE
Add support for single region 1pc

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

### CODE REVIEW

File: tikv/mvcc.go [hunk 1]
```diff
@@ -682,17 +697,23 @@ func (store *MVCCStore) prewritePessimistic(reqCtx *requestCtx, mutations []*kvr
...
 		physical, logical, tsErr := store.pdClient.GetTS(context.Background())
 		if tsErr != nil {
 			return tsErr
 		}
 		minCommitTS = uint64(physical)<<18 + uint64(logical)
-		reqCtx.asyncMinCommitTS = minCommitTS
+		if req.UseAsyncCommit {
+			reqCtx.asyncMinCommitTS = minCommitTS
+		} else if !req.TryOnePc {
+			log.Fatal("unreachable")
```
Comment (coocood) [inline tikv/mvcc.go:711]: This branch is obviously unreachable, we can just remove it.

File: tikv/mvcc.go [hunk 2]
```diff
@@ -682,17 +697,23 @@ func (store *MVCCStore) prewritePessimistic(reqCtx *requestCtx, mutations []*kvr
...
+		}
 	}
+
+	keys := make([][]byte, 0, len(mutations))
```
Comment (coocood) [inline tikv/mvcc.go:715]: Can we avoid memory allocation if not OnePC?

File: tikv/mvcc.go [hunk 3]
```diff
@@ -682,17 +697,23 @@ func (store *MVCCStore) prewritePessimistic(reqCtx *requestCtx, mutations []*kvr
...
+	locks := make([]*mvcc.MvccLock, 0, len(mutations))
 	for i, m := range mutations {
 		if m.Op == kvrpcpb.Op_CheckNotExists {
 			continue
```
Comment (coocood) [inline tikv/mvcc.go:719]: unfold the code below, we don't need to set the MinCOmmitTs in the loop.

Comment (MyonKeminta) [reply tikv/mvcc.go:719]: It looks needed for `buildPrewriteLock`.

Comment (coocood) [reply tikv/mvcc.go:719]: Yes, but we can put it out of the loop.

File: tikv/mvcc.go [hunk 4]
```diff
@@ -704,11 +725,54 @@ func (store *MVCCStore) prewriteMutations(reqCtx *requestCtx, mutations []*kvrpc
...
+	}
+	if minCommitTS < req.StartVersion {
+		log.Fatal("1pc commitTS less than startTS", zap.Uint64("startTS", req.StartVersion), zap.Uint64("minCommitTS", minCommitTS))
+	}
+
+	reqCtx.onePCCommitTS = minCommitTS
+	store.updateLatestTS(minCommitTS)
+	batch := store.dbWriter.NewWriteBatch(req.StartVersion, minCommitTS, reqCtx.rpcCtx)
+	for i, key := range keys {
+		batch.Commit(key, locks[i])
```
Comment (MyonKeminta) [inline tikv/mvcc.go:767]: There's one more problem. Trying to commit a key that is not locked will cause panicking. Looks like there need a special method for `WriteBatch` to implement it.

Comment (MyonKeminta) [reply tikv/mvcc.go:767]: I found that there's a complete raft system which saves the commit message in raftlog, which also need to be changed to support it. @coocood Do you think it's a good idea to change them all? or is there any better idea?

Comment (coocood) [reply tikv/mvcc.go:767]: We can support raft mode later.

Comment (coocood) [reply tikv/mvcc.go:767]: We can add a bool in `MvccLockHdr` for onePC.
And do not delete the lock if it is onePC.

Comment (MyonKeminta) [reply tikv/mvcc.go:767]: It sometimes needs to delete the pessimistic lock.

Comment (coocood) [reply tikv/mvcc.go:767]: We could also check if the lock is pessimistic.

**Comment by coocood** [comment]
@MyonKeminta why the diff is so large?

**Comment by MyonKeminta** [comment]
Something goes wrong. Wait a minute. Sorry

File: tikv/mvcc/mvcc.go [hunk 5]
```diff
@@ -59,6 +60,8 @@ type MvccLockHdr struct {
 	PrimaryLen     uint16
 	UseAsyncCommit bool
 	SecondaryNum   uint32
+	UseOnePC       bool
```
Comment (coocood) [inline tikv/mvcc/mvcc.go:63]: Move the two fields before `SecondaryNum` can save storage.

File: tikv/mvcc/mvcc.go [hunk 6]
```diff
@@ -59,6 +60,8 @@ type MvccLockHdr struct {
...
+	IsPessimistic  bool
```
Comment (coocood) [inline tikv/mvcc/mvcc.go:64]: We don't need this field, we can check the Op instead.

Comment (MyonKeminta) [reply tikv/mvcc/mvcc.go:64]: This happens when prewriting. Can the `Op` show whether the key is locked by a pessimistic lock? As I understood it, it only distinguishes the type of the modification like `Put`, `Delete`, `Insert`, `Lock`.

Comment (coocood) [reply tikv/mvcc/mvcc.go:64]: No, the pessimistic lock has its own code `Op_PessimisticLock Op = 5`.

**Comment by MyonKeminta** [comment]
I think maybe it's better to merge this later, since I think @sticnarf 's latest proposal may need to update kvproto again

**Comment by MyonKeminta** [comment]
kvproto was updated

File: go.mod1 [hunk 7]
```diff
@@ -12,7 +12,7 @@ require (
 	github.com/pingcap/badger v1.5.1-0.20200908111422-2e78ee155d19
 	github.com/pingcap/check v0.0.0-20200212061837-5e12011dc712
 	github.com/pingcap/errors v0.11.5-0.20200729012136-4e113ddee29e
-	github.com/pingcap/kvproto v0.0.0-20200927054727-1290113160f0
+	github.com/pingcap/kvproto master
```
Comment (coocood) [inline go.mod1:15]: Why not use a specific version?

Comment (MyonKeminta) [reply go.mod1:15]: Oh, sorry, I ran `make test` before committing and it doesn't copy go.mod back to go.mod1

**Comment by coocood** [comment]
LGTM

**Pull request merged**